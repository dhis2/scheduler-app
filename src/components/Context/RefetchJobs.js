import { createContext } from 'react'

const message =
    'RefetchJobsContext consumer needs to have a valid Provider as parent'

// Throws an error if the consumer is not nested in a provider
const RefetchJobsContext = createContext(() => {
    throw new Error(message)
})

export default RefetchJobsContext
