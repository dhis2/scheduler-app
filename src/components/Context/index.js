import RefetchJobsContext from './RefetchJobs'

export { RefetchJobsContext }
