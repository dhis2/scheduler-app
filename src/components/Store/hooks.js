import { useContext } from 'react'
import StoreContext from './StoreContext'

/**
 * The state for the job filter and showing system
 * jobs is used in the job list, but kept in the
 * store since it has to persist after a refetch.
 */

export const useJobFilter = () => {
    const store = useContext(StoreContext)
    return store.jobFilter
}

export const useShowSystemJobs = () => {
    const store = useContext(StoreContext)
    return store.showSystemJobs
}
