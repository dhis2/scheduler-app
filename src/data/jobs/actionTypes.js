export const FETCH_JOBS = 'FETCH_JOBS'
export const FETCH_JOBS_SUCCESS = 'FETCH_JOBS_SUCCESS'
export const FETCH_JOBS_FAIL = 'FETCH_JOBS_FAIL'

export const ENABLE_JOB = 'ENABLE_JOB'
export const ENABLE_JOB_SUCCESS = 'ENABLE_JOB_SUCCESS'
export const ENABLE_JOB_FAIL = 'ENABLE_JOB_FAIL'

export const DISABLE_JOB = 'DISABLE_JOB'
export const DISABLE_JOB_SUCCESS = 'DISABLE_JOB_SUCCESS'
export const DISABLE_JOB_FAIL = 'DISABLE_JOB_FAIL'

export const DELETE_JOB = 'DELETE_JOB'
export const DELETE_JOB_SUCCESS = 'DELETE_JOB_SUCCESS'
export const DELETE_JOB_FAIL = 'DELETE_JOB_FAIL'

export const RUN_JOB = 'RUN_JOB'
export const RUN_JOB_SUCCESS = 'RUN_JOB_SUCCESS'
export const RUN_JOB_FAIL = 'RUN_JOB_FAIL'
