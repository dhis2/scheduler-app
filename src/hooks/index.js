export * from './jobs'
export * from './job-schedules'
export * from './job-types'
export * from './parameter-options'
