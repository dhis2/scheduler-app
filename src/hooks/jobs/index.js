export { default as useJobById } from './use-job-by-id'
export { default as useJobs } from './use-jobs'
export { default as useSubmitJob } from './use-submit-job'
export { default as useUpdateJob } from './use-update-job'
