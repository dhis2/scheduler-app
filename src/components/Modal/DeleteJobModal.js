import React from 'react'
import { func, string } from 'prop-types'
import { Button, Modal, ButtonStrip } from '@dhis2/ui-core'
import { connect } from 'react-redux'
import { actions as modalActions } from '../../data/modal'
import { actions as jobActions } from '../../data/jobs'

export const UnconnectedDeleteJobModal = ({ id, hideModal, deleteJob }) => (
    <Modal open small onClose={hideModal}>
        <Modal.Content>Are you sure you want to delete this job?</Modal.Content>
        <Modal.Actions>
            <ButtonStrip end>
                <Button secondary name="hide-modal" onClick={hideModal}>
                    Cancel
                </Button>
                <Button
                    destructive
                    name={`delete-job-${id}`}
                    onClick={() => {
                        deleteJob(id)
                        hideModal()
                    }}
                >
                    Delete
                </Button>
            </ButtonStrip>
        </Modal.Actions>
    </Modal>
)

UnconnectedDeleteJobModal.propTypes = {
    id: string.isRequired,
    hideModal: func.isRequired,
    deleteJob: func.isRequired,
}

const mapDispatchToProps = {
    deleteJob: jobActions.deleteJob,
    hideModal: modalActions.hideModal,
}

export default connect(
    null,
    mapDispatchToProps
)(UnconnectedDeleteJobModal)
