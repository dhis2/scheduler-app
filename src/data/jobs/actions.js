import { normalize } from 'normalizr'
import urlJoin from 'url-join'
import fetchy from '../../services/fetchy'
import * as rootSelectors from '../../rootSelectors'
import endpoints from '../../services/endpoints'
import * as types from './actionTypes'
import * as schemas from './schemas'
import * as selectors from './reducer'

/**
 * Fetch jobs
 */

export const fetchJobsSuccess = payload => ({
    type: types.FETCH_JOBS_SUCCESS,
    meta: {
        receivedAt: Date.now(),
    },
    payload,
})

export const fetchJobsFail = error => ({
    type: types.FETCH_JOBS_FAIL,
    meta: {
        receivedAt: Date.now(),
    },
    error,
})

export const fetchJobs = () => dispatch => {
    dispatch({ type: types.FETCH_JOBS })

    return fetchy(urlJoin(endpoints.jobs, '?fields=*'))
        .then(data => normalize(data.jobConfigurations, [schemas.jobs]))
        .then(normalized => dispatch(fetchJobsSuccess(normalized)))
        .catch(error => dispatch(fetchJobsFail(error)))
}

export const fetchJobsIfNeeded = () => (dispatch, getState) => {
    const state = rootSelectors.getJobs(getState())
    const shouldFetch = selectors.getShouldFetch(state)

    if (!shouldFetch) {
        return Promise.resolve()
    }

    return dispatch(fetchJobs())
}

/**
 * Enable job
 */

export const enableJobSuccess = () => ({
    type: types.ENABLE_JOB_SUCCESS,
})

export const enableJobFail = error => ({
    type: types.ENABLE_JOB_FAIL,
    error,
})

export const enableJob = id => dispatch => {
    dispatch({ type: types.ENABLE_JOB })

    const fetchOptions = {
        method: 'PATCH',
        headers: {
            'content-type': 'application/json',
        },
        body: JSON.stringify({
            enabled: true,
        }),
    }

    return fetchy(urlJoin(endpoints.jobs, `/${id}`), fetchOptions)
        .then(() => dispatch(enableJobSuccess()))
        .catch(error => dispatch(enableJobFail(error)))
}

/**
 * Disable job
 */

export const disableJobSuccess = () => ({
    type: types.DISABLE_JOB_SUCCESS,
})

export const disableJobFail = error => ({
    type: types.DISABLE_JOB_FAIL,
    error,
})

export const disableJob = id => dispatch => {
    dispatch({ type: types.DISABLE_JOB })

    const fetchOptions = {
        method: 'PATCH',
        headers: {
            'content-type': 'application/json',
        },
        body: JSON.stringify({
            enabled: false,
        }),
    }

    return fetchy(urlJoin(endpoints.jobs, `/${id}`), fetchOptions)
        .then(() => dispatch(disableJobSuccess()))
        .catch(error => dispatch(disableJobFail(error)))
}

/**
 * Delete job
 */

export const deleteJobSuccess = () => ({
    type: types.DELETE_JOB_SUCCESS,
})

export const deleteJobFail = error => ({
    type: types.DELETE_JOB_FAIL,
    error,
})

export const deleteJob = id => dispatch => {
    dispatch({ type: types.DELETE_JOB })

    const fetchOptions = {
        method: 'DELETE',
    }

    return fetchy(urlJoin(endpoints.jobs, `/${id}`), fetchOptions)
        .then(() => dispatch(deleteJobSuccess()))
        .catch(error => dispatch(deleteJobFail(error)))
}

/**
 * Run job
 */

export const runJobSuccess = () => ({
    type: types.RUN_JOB_SUCCESS,
})

export const runJobFail = error => ({
    type: types.RUN_JOB_FAIL,
    error,
})

export const runJob = id => dispatch => {
    dispatch({ type: types.RUN_JOB })

    return fetchy(urlJoin(endpoints.jobs, `/${id}/execute`))
        .then(() => dispatch(runJobSuccess()))
        .catch(error => dispatch(runJobFail(error)))
}

/**
 * Create job
 */

export const createJobSuccess = () => ({
    type: types.CREATE_JOB_SUCCESS,
})

export const createJobFail = error => ({
    type: types.CREATE_JOB_FAIL,
    error,
})

export const createJob = job => dispatch => {
    const fetchOptions = {
        method: 'POST',
        headers: {
            'content-type': 'application/json',
        },
        body: JSON.stringify(job),
    }

    dispatch({ type: types.CREATE_JOB })

    return fetchy(endpoints.jobs, fetchOptions)
        .then(() => dispatch(createJobSuccess()))
        .catch(error => {
            dispatch(createJobFail(error))

            // Throw the error so final-form can catch it as well for validation
            throw error
        })
}
