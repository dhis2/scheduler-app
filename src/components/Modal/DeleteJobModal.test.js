import React from 'react'
import { shallow, mount } from 'enzyme'
import { useDeleteJob } from '../../hooks/jobs'
import { RefetchJobsContext } from '../Context'
import DeleteJobModal from './DeleteJobModal'

jest.mock('../../hooks/jobs', () => ({
    useDeleteJob: jest.fn(() => [() => {}]),
}))

describe('<DeleteJobModal>', () => {
    it('renders correctly', () => {
        const props = {
            id: 'id',
            hideModal: () => {},
        }
        const wrapper = shallow(<DeleteJobModal {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('calls hideModal when cancel button is clicked', () => {
        const props = {
            id: 'id',
            hideModal: jest.fn(),
        }
        const wrapper = mount(<DeleteJobModal {...props} />)

        wrapper.find('button[name="hide-modal"]').simulate('click')

        expect(props.hideModal).toHaveBeenCalled()
    })

    it('calls deleteJob, refetch and hideModal when delete button is clicked', async () => {
        const deletion = Promise.resolve()
        const deleteJobSpy = jest.fn(() => deletion)
        const refetchSpy = jest.fn(() => {})
        const hideModalSpy = jest.fn(() => {})
        const props = {
            id: 'id',
            hideModal: hideModalSpy,
        }

        useDeleteJob.mockImplementationOnce(() => [deleteJobSpy])

        const wrapper = mount(
            <RefetchJobsContext.Provider value={refetchSpy}>
                <DeleteJobModal {...props} />
            </RefetchJobsContext.Provider>
        )

        wrapper.find('button[name="delete-job-id"]').simulate('click')

        await deletion

        expect(deleteJobSpy).toHaveBeenCalledWith({ id: 'id' })
        expect(hideModalSpy).toHaveBeenCalled()
        expect(refetchSpy).toHaveBeenCalled()
    })

    it('calls hideModal when cover is clicked', () => {
        const props = {
            id: 'id',
            hideModal: jest.fn(),
        }
        const wrapper = mount(<DeleteJobModal {...props} />)

        wrapper.find('div[data-test="dhis2-uicore-layer"]').simulate('click')

        expect(props.hideModal).toHaveBeenCalled()
    })
})
