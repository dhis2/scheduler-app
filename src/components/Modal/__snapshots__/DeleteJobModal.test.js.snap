// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteJobModal> renders correctly 1`] = `
<Modal
  onClose={[Function]}
  open={true}
  small={true}
>
  <Content>
    Are you sure you want to delete this job?
  </Content>
  <Actions>
    <ButtonStrip
      end={true}
    >
      <Button
        name="hide-modal"
        onClick={[Function]}
        secondary={true}
        type="button"
      >
        Cancel
      </Button>
      <Button
        destructive={true}
        name="delete-job-id"
        onClick={[Function]}
        type="button"
      >
        Delete
      </Button>
    </ButtonStrip>
  </Actions>
</Modal>
`;
